### PR TITLE
Improve docstring for names()

### LINKED ISSUE #2747: Allow selecting columns based on predicate on column contents
Based on a discussion from slack:
> [Sebastian Rollén]: Potentially dumb question: How might I `select` only columns that don't have missing data in them? In R/dplyr I would use select_if, so I was looking for something similar
> 
> [Sebastian Rollén]: my best approach so far is to choose the column indices with `indices = map(x -> !any(ismissing, x), eachcol(df))` and then `df[:, indices]` , but I feel like there's probably a nicer solution?
>
> [Peter Deffebach]: you can use a type in `names`, something like `names(df, <:Missing)`. But this is all columns that allow missings, not the ones with missings.
>
> [Sebastian Rollén]: hm, that probably won't work in this case as all my columns allow missings as they are `unstack`ed from the same column
>
> [Sebastian Rollén]: I'm explicitly looking for the columns that don't have any missing values so I can filter out the other ones and then run `disallowmissing` so that I can get rid of the type union
>
> [Bogumił Kamiński]: What you use now is a currently recommended way. Currently in `names` if you pass predicate it looks at column name not at column contents. We have decided for this as it seemed more common. But agreed it is also useful to find columns based on a predicate of their contents. Maybe open an issue and we can discuss there how to design it.

It's currently possible to `select` a dataframe using a predicate in `names`, but this predicate applies to the types of the columns, not the contents of the columns themselves. The idea here is to have some function that allows `select`ing a column based on the contents of the columns instead, similar to `select_if` in dplyr

### PR DESCRIPTION
**Pull request by xluo127**
Hi, @bkamins! Based on the comments in #2877, improving docstring rather than adding a new argument should be better. Examples that might be able to fix #2747 are added.

If there is something wrong or not suitable, please let me know, and I will manage to fix them.

### CODE REVIEW

File: docs/src/man/getting_started.md [hunk 1]
```diff
@@ -127,8 +127,25 @@ julia> names(df)
  "B"
 ```
 
-To get column names as `Symbol`s use the `propertynames` function:
+You can also extract column names by regular expression, indices or `Symbol`s:
```
Comment (bkamins) [inline docs/src/man/getting_started.md:130]:
```suggestion
You can also filter column names by column selector condition that is passed as a second argument.
See [`names`](@ref) docstring for a detailed list of available conditions.
Here we give some selected examples:
```

File: docs/src/man/getting_started.md [hunk 2]
```diff
@@ -127,8 +127,25 @@ julia> names(df)
...
+
+```jldoctest dataframe
+julia> names(df, r"A")
```
Comment (bkamins) [inline docs/src/man/getting_started.md:133]:
```suggestion
julia> names(df, r"A") # a regular expression selector
```

File: docs/src/man/getting_started.md [hunk 3]
```diff
@@ -127,8 +127,25 @@ julia> names(df)
...
+1-element Vector{String}:
+ "A"
+
+julia> names(df, 1)
```
Comment (bkamins) [inline docs/src/man/getting_started.md:137]:
```suggestion
julia> names(df, Int) # a selector using column element type
```

File: docs/src/man/getting_started.md [hunk 4]
```diff
@@ -127,8 +127,25 @@ julia> names(df)
...
+1-element Vector{String}:
+ "A"
+
+julia> names(df, :A)
```
Comment (bkamins) [inline docs/src/man/getting_started.md:141]:
```suggestion
julia> names(df, Not(:B)) # selector keeping all columns except :B
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 5]
```diff
@@ -63,9 +63,16 @@ abstract type AbstractDataFrame end
 """
     names(df::AbstractDataFrame)
     names(df::AbstractDataFrame, cols)
+    names(df::GroupDataFrame, cols)
+    names(df::GroupKey)
+    names(df::DataFrameRows, cols)
+    names(df::DataFrameRow, cols)
+    names(df::DataFrameColumns, cols)
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:70]:
```suggestion
    names(df::AbstractDataFrame, cols=:)
    names(df::DataFrameRow, cols=:)
    names(df::GroupDataFrame, cols=:)
    names(df::DataFrameRows, cols=:)
    names(df::DataFrameColumns, cols=:)
    names(df::GroupKey)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 6]
```diff
@@ -61,11 +61,17 @@ abstract type AbstractDataFrame end
...
+    names(df::AbstractDataFrame, cols=:)
+    names(df::DataFrameRow, cols=:)
+    names(df::GroupDataFrame, cols=:)
+    names(df::DataFrameRows, cols=:)
+    names(df::DataFrameColumns, cols=:)
+    names(df::GroupKey)
 
 Return a freshly allocated `Vector{String}` of names of columns contained in `df`.
 
+`df` can be some type other than `AbstractDataFrame`, but do not call `names(df)` for a `GroupKeys`.
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:73]:
```suggestion
For all supported types of `df` except `GroupKeys` you can optionally pass a `cols` positional argument.
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 7]
```diff
@@ -75,7 +81,61 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
 * a `Function` predicate taking the column name as a string and returning `true`
   for columns that should be kept
 
+Note that for a `GroupKey`, `cols` cannot be passed.
+
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:85]:
```suggestion
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 8]
```diff
@@ -76,6 +82,58 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+   3 │ missing       4        2       4
+
+julia> disallowmissing!(df, 2)
+3×4 DataFrame
+ Row │ x1       x2     x3       x4
+     │ Int64?   Int64  Int64?   Int64?
+─────┼─────────────────────────────────
+   1 │       1      3        3       2
+   2 │ missing      2  missing       4
+   3 │ missing      4        2       4
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:104]:
```suggestion
julia> df = DataFrame([1 3 3 2; missing 2 missing 4; missing 4 2 4], :auto);

julia> disallowmissing!(df, 2)
3×4 DataFrame
 Row │ x1       x2     x3       x4
     │ Int64?   Int64  Int64?   Int64?
─────┼─────────────────────────────────
   1 │       1      3        3       2
   2 │ missing      2  missing       4
   3 │ missing      4        2       4
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 9]
```diff
@@ -76,6 +82,51 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:98]:
```suggestion

julia> names(df)
4-element Vector{String}:
 "x1"
 "x2"
 "x3"
 "x4"
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 10]
```diff
@@ -76,6 +82,57 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+julia> names(df)
+4-element Vector{String}:
+ "x1"
+ "x2"
+ "x3"
+ "x4"
+julia> names(df, Int64)
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:105]:
```suggestion
julia> names(df, Int64)
```
```suggestion

julia> names(df, Int) # pick columns whose element type is Int
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 11]
```diff
@@ -76,6 +82,58 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+ "x1"
+ "x2"
+ "x3"
+ "x4"
+
+julia> names(df, Int) # pick columns whose element type is Int
+1-element Vector{String}:
+ "x2"
+
+julia> names(df, x -> x[2] == '2')
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:110]:
```suggestion
julia> names(df, x -> x[end] == '2') # pick columns for which last character in their name is '2'
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 12]
```diff
@@ -76,6 +82,58 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+1-element Vector{String}:
+ "x2"
+
+julia> fun(col) = sum(skipmissing(col)) >= 10
+fun (generic function with 1 method)
+
+julia> names(df, fun.(eachcol(df)))
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:117]:
```suggestion
julia> names(df, fun.(eachcol(df))) # pick columns for which sum of their elements is at least 10
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 13]
```diff
@@ -76,6 +82,58 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+1-element Vector{String}:
+ "x4"
+
+julia> names(df, .>:(eltype.(eachcol(df)), Missing))
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:121]:
```suggestion
julia> names(df, .>:(eltype.(eachcol(df)), Missing)) # pick columns that allow missing values
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 14]
```diff
@@ -76,6 +82,58 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+ "x1"
+ "x3"
+ "x4"
+
+julia> names(df, [any(map(ismissing, col)) for col in eachcol(df)])
+2-element Vector{String}:
+ "x1"
+ "x3"
+
+julia> names(df, any.(ismissing, eachcol(df)))
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:132]:
```suggestion
julia> names(df, any.(ismissing, eachcol(df))) # pick columns that contain missing values
```

**Comment by bkamins** [comment]
Thank you for the PR. I have applied some changes so that you can see what things we would typically do.

@nalimilan - as usual could you please have a look at the PR (and especially the wording)? Thank you!

**Comment by xluo127** [comment]
Thanks for guiding! I will see the changes and learn.

File: docs/src/man/getting_started.md [hunk 15]
```diff
@@ -127,8 +127,27 @@ julia> names(df)
  "B"
 ```
 
-To get column names as `Symbol`s use the `propertynames` function:
+You can also filter column names by column selector condition that is passed as a second argument.
+See [`names`](@ref) docstring for a detailed list of available conditions.
```
Comment (nalimilan) [inline docs/src/man/getting_started.md:131]:
```suggestion
You can also filter column names by passing a column selector condition as a second argument.
See the [`names`](@ref) docstring for a detailed list of available conditions.
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 16]
```diff
@@ -61,11 +61,17 @@ abstract type AbstractDataFrame end
...
+    names(df::DataFrameRow, cols=:)
+    names(df::GroupDataFrame, cols=:)
+    names(df::DataFrameRows, cols=:)
+    names(df::DataFrameColumns, cols=:)
+    names(df::GroupKey)
 
 Return a freshly allocated `Vector{String}` of names of columns contained in `df`.
 
+For all supported types of `df` except `GroupKeys` you can optionally pass a `cols` positional argument.
+
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:74]: This is already visible in the signatures, so I wouldn't repeat it.
```suggestion
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 17]
```diff
@@ -61,11 +61,17 @@ abstract type AbstractDataFrame end
 ##############################################################################
 
 """
-    names(df::AbstractDataFrame)
-    names(df::AbstractDataFrame, cols)
+    names(df::AbstractDataFrame, cols=:)
+    names(df::DataFrameRow, cols=:)
+    names(df::GroupDataFrame, cols=:)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:66]:
```suggestion
    names(df::GroupedDataFrame, cols=:)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 18]
```diff
@@ -76,6 +82,53 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
   for columns that should be kept
 
 See also [`propertynames`](@ref) which returns a `Vector{Symbol}`.
+
+# Examples
+```jldoctest
+julia> df = DataFrame([1 3 3 2; missing 2 missing 4; missing 4 2 4], :auto);
+
+julia> disallowmissing!(df, 2)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:90]: This looks simpler to me:
```suggestion
julia> df = DataFrame(x1=[1, missing, missing], x2=[3, 2, 4], x3=[3, missing, 2], x4=[2, 4, 4])
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 19]
```diff
@@ -76,6 +82,53 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
...
+ "x2"
+
+julia> fun(col) = sum(skipmissing(col)) >= 10
+fun (generic function with 1 method)
+
+julia> names(df, fun.(eachcol(df))) # pick columns for which sum of their elements is at least 10
+1-element Vector{String}:
+ "x4"
+
+julia> names(df, .>:(eltype.(eachcol(df)), Missing)) # pick columns that allow missing values
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:121]: This works, right?
```suggestion
julia> names(df, eltype.(eachcol(df)) .>: Missing) # pick columns that allow missing values
```

**Comment by bkamins** [comment]
Thank you. I will merge this when CI passes.

File: src/abstractdataframe/abstractdataframe.jl [hunk 20]
```diff
@@ -76,6 +80,51 @@ selector (this is useful in particular with regular expressions, `Cols`, `Not`,
   for columns that should be kept
 
 See also [`propertynames`](@ref) which returns a `Vector{Symbol}`.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(x1=[1, missing, missing], x2=[3, 2, 4], x3=[3, missing, 2], x4=[2, 4, 4])
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:86]:
```suggestion
julia> df = DataFrame(x1=[1, missing, missing], x2=[3, 2, 4], x3=[3, missing, 2], x4=Union{Int, Missing}[2, 4, 4])
```

**Comment by bkamins** [comment]
@nalimilan - OK to merge?

**Review by bkamins — APPROVED** [review]

**Pull request merged**